Source/WebCore/dom/TreeScope.cpp

```diff
@@ -553,19 +548,33 @@ RadioButtonGroups& TreeScope::radioButtonGroups()
     return *m_radioButtonGroups;
 }
 
+CSSStyleSheetObservableArray& TreeScope::ensureAdoptedStyleSheets()
+{
+    if (UNLIKELY(!m_adoptedStyleSheets)) {
+        if (auto* shadowRoot = dynamicDowncast<ShadowRoot>(m_rootNode))
+            m_adoptedStyleSheets = CSSStyleSheetObservableArray::create(*shadowRoot);
+        else
+            m_adoptedStyleSheets = CSSStyleSheetObservableArray::create(downcast<Document>(m_rootNode));
+    }
+    return *m_adoptedStyleSheets;
+}
+
 const Vector<RefPtr<CSSStyleSheet>>& TreeScope::adoptedStyleSheets() const
 {
-    return m_adoptedStyleSheets->sheets();
+    static NeverDestroyed<Vector<RefPtr<CSSStyleSheet>>> emptyStyleSheets;
+    return m_adoptedStyleSheets ? m_adoptedStyleSheets->sheets() : emptyStyleSheets.get();
```

Comment:
If I was adding this function now I would return a `span<const RefPtr<CSSStyleSheet>>` rather than a `const&` or expose a `forEach` function.